torchok/constructor/constructor.py

```diff
@@ -80,14 +83,174 @@ def configure_optimizers(self, parameters: Union[Module, Tensor, List[Union[Modu
+          rate for all bias parameters (except for those in normalization
+          layers and offset layers of DCN).
+        - ``bias_decay_mult`` (float): It will be multiplied to the weight
+          decay for all bias parameters (except for those in
+          normalization layers, depthwise conv layers, offset layers of DCN).
+        - ``norm_decay_mult`` (float): It will be multiplied to the weight
+          decay for all weight and bias parameters of normalization
+          layers.
+        - ``dwconv_decay_mult`` (float): It will be multiplied to the weight
+          decay for all weight and bias parameters of depthwise conv
+          layers.
+        - ``dcn_offset_lr_mult`` (float): It will be multiplied to the learning
+          rate for parameters of offset layer in the deformable convs
+          of a model.
+
+        Note:
+
+            1. If the option ``dcn_offset_lr_mult`` is used, the constructor will
+            override the effect of ``bias_lr_mult`` in the bias of offset layer.
+            So be careful when using both ``bias_lr_mult`` and
+            ``dcn_offset_lr_mult``. If you wish to apply both of them to the offset
+            layer in deformable convs, set ``dcn_offset_lr_mult`` to the original
+            ``dcn_offset_lr_mult`` * ``bias_lr_mult``.
+
+            2. If the option ``dcn_offset_lr_mult`` is used, the constructor will
+            apply it to all the DCN layers in the model. So be careful when the
+            model contains multiple DCN layers in places other than backbone.
+
+        Args:
+            modules: The module or list of modules with parameters to be optimized.
+            optimizer_params: The config dict of the optimizer.
+                Positional fields are
+                    - `name`: class name of the optimizer.
+
+                Optional fields are
+                    - `params`: dict with parameters to set up an optimize, e.g., lr, weight_decay, momentum.
+                    - `paramwise_cfg`:
+        """
+
         optimizer_class = OPTIMIZERS.get(optimizer_params.name)
-        parameters = Constructor._set_weight_decay_for_parameters(parameters)
-        optimizer = optimizer_class(parameters, **optimizer_params.params)
+        paramwise_cfg = optimizer_params.paramwise_cfg
+        optimizer_cfg = optimizer_params.params
+
+        if isinstance(modules, (tuple, list)):
+            modules = ModuleList(modules)
+
+        # parameters: List[Union[Dict, Parameter]] = list(modules.parameters())
+        if not paramwise_cfg:
+            parameters: List[Union[Dict, Parameter]] = list(modules.parameters())
+        else:
+            # set param-wise lr and weight decay recursively
+            parameters: List[Union[Dict, Parameter]] = []
+            Constructor.add_params(parameters, modules, optimizer_cfg, paramwise_cfg)
+
+        optimizer = optimizer_class(parameters, **optimizer_cfg)
```

Comment:
Ok, one question. In Swin backbone we have `no_weight_decay()` method which return parameters which `weights_decay` should equal 0. But now this method is not calls in `create_optimizer()`. So now we need to write `weights_decay = 0` in config?